# opening phyloseq objects
#OTU table (shared file)

library(phyloseq)

OTU = read.table("example.final.an.unique_list.0.03.norm.shared", header=TRUE, sep="\t")

#Taxonomy of each OTU
tax = read.table("example.final.an.unique_list.0.03.cons.taxonomy", header=TRUE, sep="\t")

#Metadata. Since we made this in Excel, not mothur, we can use the "row.names" modifier to automatically name the rows by the values in the first column (sample names)
meta = read.table("example.metadata.txt", header=TRUE, row.names=1, sep="\t")


library(microbiome)

# Import output CSV files generated by write_phyloseq
pseq1 <- read_phyloseq(otu.file = , taxonomy.file = , metadata.file = , type = "simple")

# Import mother .shared and .taxonomy and metadata files
pseq2 <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = "mothur")

# Import BIOM files
pseq3 <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = "biom")


######## import biom file

biom_file <- paste(uzdir, "study_1011_closed_reference_otu_table.biom", sep = "")
map_file <- paste(uzdir, "study_1011_mapping_file.txt", sep = "")
# Now import the .biom-formatted otu_table-tax_table file.
biom_otu_tax <- import_biom(biom_file, "greengenes")
# Add sample data to the dataset using merge
physeq <- import_qiime_sample_data(map_file)
class(physeq)

colnames(tax_table(physeq)) = c("Root", "Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")



##### import qiime
# qiimedata is the phyloseq object

otufile = system.file("extdata", "GP_otu_table_rand_short.txt.gz", package="phyloseq")
mapfile = system.file("extdata", "master_map.txt", package="phyloseq")
trefile = system.file("extdata", "GP_tree_rand_short.newick.gz", package="phyloseq")
rs_file = system.file("extdata", "qiime500-refseq.fasta", package="phyloseq")
qiimedata = import_qiime(otufile, mapfile, trefile, rs_file)


##### microbiomeSeq
OTU = otu_table(as.matrix(abund_table), taxa_are_rows = FALSE)
TAX = tax_table(as.matrix(OTU_taxonomy))
SAM = sample_data(meta_table)
OTU_tree <- compute.brlen(OTU_tree, method = "Grafen")
physeq <- merge_phyloseq(phyloseq(OTU, TAX), SAM, OTU_tree)




otu <- as.matrix(read.table("data/mach/otu_table.tsv"))
tax <- as.matrix(read.table("data/mach/tax_table.tsv"))
tree <- read.tree("data/mach/tree.nwk")
map <- read.table("data/mach/metadata.tsv")